## Week 3 Quiz

***

## Project Description

Use the dognition_data_no_aggregation data set provided in this course for this quiz.

### Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from statsmodels.formula.api import ols
import datetime
from datetime import datetime, timedelta
import scipy.stats
import pandas_profiling
from pandas_profiling import ProfileReport


%matplotlib inline
#sets the default autosave frequency in seconds
%autosave 60 
sns.set_style('dark')
sns.set(font_scale=1.2)

plt.rc('axes', titlesize=9)
plt.rc('axes', labelsize=14)
plt.rc('xtick', labelsize=12)
plt.rc('ytick', labelsize=12)

import warnings
warnings.filterwarnings('ignore')

# Use Folium library to plot values on a map.
#import folium

# Use Feature-Engine library
#import feature_engine
#import feature_engine.missing_data_imputers as mdi
#from feature_engine.outlier_removers import Winsorizer
#from feature_engine import categorical_encoders as ce
#from feature_engine.discretisation import EqualWidthDiscretiser, EqualFrequencyDiscretiser, DecisionTreeDiscretiser
#from feature_engine.encoding import OrdinalEncoder

pd.set_option('display.max_columns',None)
#pd.set_option('display.max_rows',None)
pd.set_option('display.width', 1000)
pd.option_context('float_format','{:.2f}'.format)

np.random.seed(0)
np.set_printoptions(suppress=True)

Autosaving every 60 seconds


In [2]:
df = pd.read_csv("dognition_data_no_aggregation.csv")

In [3]:
df

Created_at       Updated_at                               User_ID                                Dog_ID               Test_name Subcategory_name Gender  Birthday     Breed                        Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed       City State     Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID
0       13/5/2013 16:17  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Warm-up          Empathy   male      2009  Shih Tzu                        Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1     Durham    NC   27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0
1       22/5/2013 18:54  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b               Yawn Game          Empathy   male      2009  Shih Tzu                        Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1     Durham    NC   27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0
2       22/5/2013 18:55  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Warm-up          Empathy   male      2009  Shih Tzu                        Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1     Durham    NC   27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0
3       22/5/2013 18:56  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Eye Contact Game          Empathy   male      2009  Shih Tzu                        Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1     Durham    NC   27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0
4       22/5/2013 18:58  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b           Treat Warm-up    Communication   male      2009  Shih Tzu                        Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1     Durham    NC   27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0
...                 ...              ...                                   ...                                   ...                     ...              ...    ...       ...       ...                               ...         ...     ...        ...         ...       ...            ...       ...            ...         ...        ...   ...     ...     ...     ...             ...              ...             ...    ...             ...            ...
177476   14/5/2013 8:45   14/5/2013 8:45  ce2777e6-7144-11e5-ba71-058fbc01cf0b  fd444202-7144-11e5-ba71-058fbc01cf0b         One Cup Warm-up           Memory   male      2012     Mixed  Mixed Breed/ Other/ I Don't Know         NaN      50        1.0         0.0       NaN             12         1            1.0           1  Singapore   NaN  466311      SG     NaN             NaN   28/4/2014 1:18               1    NaN            12.0           12.0
177477   14/5/2013 8:50   14/5/2013 8:50  ce2777e6-7144-11e5-ba71-058fbc01cf0b  fd444202-7144-11e5-ba71-058fbc01cf0b         Two Cup Warm-up           Memory   male      2012     Mixed  Mixed Breed/ Other/ I Don't Know         NaN      50        1.0         0.0       NaN             12         1            1.0           1  Singapore   NaN  466311      SG     NaN             NaN   28/4/2014 1:18

## Exploratory Data Analysis

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177481 entries, 0 to 177480
Data columns (total 30 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Created_at        177481 non-null  object 
 1   Updated_at        177481 non-null  object 
 2   User_ID           177481 non-null  object 
 3   Dog_ID            177481 non-null  object 
 4   Test_name         177481 non-null  object 
 5   Subcategory_name  177481 non-null  object 
 6   Gender            177481 non-null  object 
 7   Birthday          177481 non-null  int64  
 8   Breed             177481 non-null  object 
 9   Breed_Type        177481 non-null  object 
 10  Breed_Group       88923 non-null   object 
 11  Weight            177481 non-null  int64  
 12  Dog_Fixed         175317 non-null  float64
 13  DNA_Tested        52968 non-null   float64
 14  Dimension         91070 non-null   object 
 15  Sign_in_Count     177481 non-null  int64  
 16  Max_Dogs          17

In [5]:
df.describe()

Birthday         Weight      Dog_Fixed    DNA_Tested  Sign_in_Count       Max_Dogs  Membership_ID     Subscribed  Rank_by_UserID  Rank_by_DogID
count  177481.000000  177481.000000  175317.000000  52968.000000  177481.000000  177481.000000   176335.00000  177481.000000   177025.000000  177025.000000
mean     2009.898265      42.683442       0.876144      0.141633      11.441771       1.373251        7.11325       0.665192       14.008728       8.494088
std         3.280318      29.171277       0.329418      0.348676      69.452489       1.545094        4.20112       0.471925       49.376213       6.670537
min      1982.000000       0.000000       0.000000      0.000000       0.000000       0.000000        1.00000       0.000000        1.000000       1.000000
25%      2008.000000      20.000000       1.000000      0.000000       3.000000       1.000000        2.00000       0.000000        3.000000       3.000000
50%      2011.000000      40.000000       1.000000      0.000000       5.000000       1.000000       10.00000       1.000000        8.000000       7.000000
75%      2012.000000      60.000000       1.000000      0.000000      10.000000       1.000000       10.00000       1.000000       15.000000      13.000000
max      2015.000000     190.000000       1.000000      1.000000    1695.000000      36.000000       17.00000       1.000000     1075.000000      45.000000

In [6]:
df.columns

Index(['Created_at', 'Updated_at', 'User_ID', 'Dog_ID', 'Test_name', 'Subcategory_name', 'Gender', 'Birthday', 'Breed', 'Breed_Type', 'Breed_Group', 'Weight', 'Dog_Fixed', 'DNA_Tested', 'Dimension', 'Sign_in_Count', 'Max_Dogs', 'Membership_ID', 'Subscribed', 'City', 'State', 'Zip', 'Country', 'Exclude', 'Free_Start_User', 'Last_Active_At', 'Membership_Type', 'Rating', 'Rank_by_UserID', 'Rank_by_DogID'], dtype='object')

### Groupby Function

In [7]:
df.groupby(["Test_name","Subcategory_name"])["Exclude"].count()

Test_name                      Subcategory_name  
1 vs 1 Game                    Social Bias            43
3 vs 1 Game                    Numerosity             60
5 vs 1 Game                    Numerosity             61
                               Social Bias            40
Arm Pointing                   Communication         600
Cover Your Eyes                Cunning               498
Delayed Cup Game               Memory                428
Different Perspective          Perspective Game       21
Expression Game                Expression Game        25
Eye Contact Game               Empathy               695
Eye Contact Warm-up            Empathy               709
Foot Pointing                  Communication         566
Impossible Task Game           Impossible Task        78
Impossible Task Warm-up        Impossible Task        81
Inferential Reasoning Game     Reasoning             393
Inferential Reasoning Warm-up  Reasoning             398
Memory versus Pointing         Memory 

In the dognition_data_no_aggregation data set, after which game is the drop-off of completed tests the greatest?

In [8]:
df.groupby(["Test_name","Subcategory_name"])["Exclude"].count().max()

775

If you remove all the entries associated with Shih Tzu dogs that weigh 190 pounds from the dognition_data_no_aggregation data set, how many different sequences of tests were administered to customers in the data that are left?

In [9]:
noshihtzu = df[df["Breed"] != "Shih Tzu"]

In [10]:
noshihtzu.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday                                              Breed   Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed          City State    Zip Country Exclude Free_Start_User Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID
20  17/5/2013 17:45  17/5/2013 17:45  ce134be0-7144-11e5-ba71-058fbc01cf0b  fd27c956-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2010  German Shepherd Dog-Nova Scotia Duck Tolling R...  Cross Breed         NaN      30        1.0         0.0       NaN             65         3            2.0           1  Hillsborough    NC  27278      US     NaN             NaN            NaN               2    NaN             1.0            1.0
21  17/5/2013 17:50  17/5/2013 17:50  ce134be0-7144-11e5-ba71-058fbc01cf0b  fd27c956-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2010  German Shepherd Dog-Nova Scotia Duck Tolling R...  Cross Breed         NaN      30        1.0         0.0       NaN             65         3            2.0           1  Hillsborough    NC  27278      US     NaN             NaN            NaN               2    NaN             2.0            2.0
22  17/5/2013 17:55  17/5/2013 17:55  ce134be0-7144-11e5-ba71-058fbc01cf0b  fd27c956-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2010  German Shepherd Dog-Nova Scotia Duck Tolling R...  Cross Breed         NaN      30        1.0         0.0       NaN             65         3            2.0           1  Hillsborough    NC  27278      US     NaN             NaN            NaN               2    NaN             3.0            3.0
23  17/5/2013 17:57  17/5/2013 17:57  ce134be0-7144-11e5-ba71-058fbc01cf0b  fd27c956-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2010  German Shepherd Dog-Nova Scotia Duck Tolling R...  Cross Breed         NaN      30        1.0         0.0       NaN             65         3            2.0           1  Hillsborough    NC  27278      US     NaN             NaN            NaN               2    NaN             4.0            4.0
24  17/5/2013 18:02  17/5/2013 18:02  ce134be0-7144-11e5-ba71-058fbc01cf0b  fd27c956-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2010  German Shepherd Dog-Nova Scotia Duck Tolling R...  Cross Breed         NaN      30        1.0         0.0       NaN             65         3            2.0           1  Hillsborough    NC  27278      US     NaN             NaN            NaN               2    NaN             5.0            5.0

In [11]:
noshihtzu["Test_name"].nunique()

41

During which day of the week do customers play Dognition games the most?

In [12]:
df["Created_at"] = pd.to_datetime(df["Created_at"])

In [13]:
df.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday     Breed  Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed    City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID
0 2013-05-13 16:17:00  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0
1 2013-05-22 18:54:00  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0
2 2013-05-22 18:55:00  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0
3 2013-05-22 18:56:00  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0
4 2013-05-22 18:58:00  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0

In [14]:
#Return the day of the week as an integer, where Monday is 0 and Sunday is 6.
df["WeekDay"] = df["Created_at"].dt.weekday

In [15]:
df.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday     Breed  Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed    City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID  WeekDay
0 2013-05-13 16:17:00  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0        0
1 2013-05-22 18:54:00  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0        2
2 2013-05-22 18:55:00  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0        2
3 2013-05-22 18:56:00  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0        2
4 2013-05-22 18:58:00  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0        2

In [16]:
df["WeekDay"].value_counts()

6    29458
1    27174
2    26216
0    25957
3    23756
5    22629
4    22291
Name: WeekDay, dtype: int64

During which month in the data set were the most tests completed?

In [17]:
df["Month"] = df["Created_at"].dt.month

In [18]:
df["Year"] = df["Created_at"].dt.year

In [19]:
df.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday     Breed  Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed    City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID  WeekDay  Month  Year
0 2013-05-13 16:17:00  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0        0      5  2013
1 2013-05-22 18:54:00  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0        2      5  2013
2 2013-05-22 18:55:00  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0        2      5  2013
3 2013-05-22 18:56:00  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0        2      5  2013
4 2013-05-22 18:58:00  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0        2      5  2013

In [20]:
df.groupby(["Month","Year"])['Created_at'].count()

Month  Year
1      2013     1914
       2014     3137
       2015     5360
2      2013     4197
       2014     2559
       2015     3882
3      2013     3752
       2014     2942
       2015     3826
4      2013     3963
       2014     2453
       2015     3212
5      2013     4417
       2014     2592
       2015    12841
6      2013     3268
       2014     5156
       2015    13607
7      2013     4398
       2014     4912
       2015     6039
8      2013     5262
       2014     4452
       2015     4481
9      2013     3568
       2014     4508
       2015    10368
10     2013     4196
       2014    15762
       2015     1440
11     2013     3461
       2014     7261
       2015     1785
12     2013     3542
       2014     7335
       2015     1633
Name: Created_at, dtype: int64

After adjusting the time stamps of the tests completed by United States users provided in the “Created At” field of the dognition_data_no_aggregation data set for time zone differences, during which hour of the day do Dognition customers play the most amount of games?

In [21]:
usausers = df[df["Country"] == "US"]

In [22]:
usausers.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday     Breed  Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed    City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID  WeekDay  Month  Year
0 2013-05-13 16:17:00  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0        0      5  2013
1 2013-05-22 18:54:00  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0        2      5  2013
2 2013-05-22 18:55:00  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0        2      5  2013
3 2013-05-22 18:56:00  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0        2      5  2013
4 2013-05-22 18:58:00  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0        2      5  2013

In [23]:
usausers["Time"] = usausers["Created_at"].dt.hour

In [24]:
usausers.head()

Created_at       Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday     Breed  Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed    City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID  WeekDay  Month  Year  Time
0 2013-05-13 16:17:00  13/5/2013 16:17  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             1.0            1.0        0      5  2013    16
1 2013-05-22 18:54:00  22/5/2013 18:54  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             2.0            2.0        2      5  2013    18
2 2013-05-22 18:55:00  22/5/2013 18:55  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             3.0            3.0        2      5  2013    18
3 2013-05-22 18:56:00  22/5/2013 18:56  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             4.0            4.0        2      5  2013    18
4 2013-05-22 18:58:00  22/5/2013 18:58  ce134a78-7144-11e5-ba71-058fbc01cf0b  fd3d1b44-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2009  Shih Tzu  Pure Breed         Toy     190        1.0         0.0  einstein            310         3            2.0           1  Durham    NC  27701      US       1               0  23/2/2015 13:39               2    NaN             5.0            5.0        2      5  2013    18

In [25]:
usausers["Time"].value_counts()

0     12582
1     11791
23    10785
22     8983
2      8716
21     8623
20     8431
19     7766
18     7283
17     6382
3      5610
16     5493
15     4933
14     3700
4      3291
13     2200
5      1545
12     1061
6       675
11      400
7       364
10      186
8       175
9       104
Name: Time, dtype: int64

In the dognition_data_no_aggregation data set, approximately what percentage of users who begin taking the Dognition Assessment complete 20 tests?

In [26]:
usertests = df.groupby("User_ID")["Created_at","Dog_ID"].count()
usertests

Created_at  Dog_ID
User_ID                                                 
ce134492-7144-11e5-ba71-058fbc01cf0b          18      18
ce134a78-7144-11e5-ba71-058fbc01cf0b          20      20
ce134be0-7144-11e5-ba71-058fbc01cf0b          11      11
ce134d16-7144-11e5-ba71-058fbc01cf0b           4       4
ce134e42-7144-11e5-ba71-058fbc01cf0b          41      41
...                                          ...     ...
ce9b2aba-7144-11e5-ba71-058fbc01cf0b           7       7
ce9b2eb6-7144-11e5-ba71-058fbc01cf0b           1       1
ce9b306e-7144-11e5-ba71-058fbc01cf0b          20      20
ce9b33b6-7144-11e5-ba71-058fbc01cf0b           4       4
ce9b35e6-7144-11e5-ba71-058fbc01cf0b           6       6

[16261 rows x 2 columns]

In [27]:
usertests["Dog_ID"].value_counts()

4      3397
20     2615
7      2026
2      1253
1      1032
       ... 
162       1
130       1
90        1
82        1
99        1
Name: Dog_ID, Length: 94, dtype: int64

In [28]:
usertests["Dog_ID"].sum()

177481

In [29]:
print("Percentage:" , 2615/177481 * 100)

Percentage: 1.4733971523712397


In the dognition_data_no_aggregation data set, what percentage of users who begin taking the Dognition Assessment using a “Free Start” promotion complete 20 tests?

In [30]:
freestartusers = df[df["Free_Start_User"] == 1]

In [31]:
freestartusers.head()

Created_at      Updated_at                               User_ID                                Dog_ID            Test_name Subcategory_name Gender  Birthday  Breed                        Breed_Type Breed_Group  Weight  Dog_Fixed  DNA_Tested Dimension  Sign_in_Count  Max_Dogs  Membership_ID  Subscribed        City State    Zip Country Exclude Free_Start_User   Last_Active_At Membership_Type Rating  Rank_by_UserID  Rank_by_DogID  WeekDay  Month  Year
1268 2013-09-06 18:20:00  9/6/2013 18:20  ce28c2cc-7144-11e5-ba71-058fbc01cf0b  fd45d3ba-7144-11e5-ba71-058fbc01cf0b         Yawn Warm-up          Empathy   male      2012  Mixed  Mixed Breed/ Other/ I Don't Know         NaN      30        1.0         0.0   charmer             30         4            2.0           1  Wilmington    DE  19802      US       0               1  3/10/2014 16:18               2    NaN             1.0            1.0        4      9  2013
1269 2013-09-06 18:24:00  9/6/2013 18:24  ce28c2cc-7144-11e5-ba71-058fbc01cf0b  fd45d3ba-7144-11e5-ba71-058fbc01cf0b            Yawn Game          Empathy   male      2012  Mixed  Mixed Breed/ Other/ I Don't Know         NaN      30        1.0         0.0   charmer             30         4            2.0           1  Wilmington    DE  19802      US       0               1  3/10/2014 16:18               2    NaN             2.0            2.0        4      9  2013
1270 2013-09-06 18:27:00  9/6/2013 18:27  ce28c2cc-7144-11e5-ba71-058fbc01cf0b  fd45d3ba-7144-11e5-ba71-058fbc01cf0b  Eye Contact Warm-up          Empathy   male      2012  Mixed  Mixed Breed/ Other/ I Don't Know         NaN      30        1.0         0.0   charmer             30         4            2.0           1  Wilmington    DE  19802      US       0               1  3/10/2014 16:18               2    NaN             3.0            3.0        4      9  2013
1271 2013-09-06 18:33:00  9/6/2013 18:33  ce28c2cc-7144-11e5-ba71-058fbc01cf0b  fd45d3ba-7144-11e5-ba71-058fbc01cf0b     Eye Contact Game          Empathy   male      2012  Mixed  Mixed Breed/ Other/ I Don't Know         NaN      30        1.0         0.0   charmer             30         4            2.0           1  Wilmington    DE  19802      US       0               1  3/10/2014 16:18               2    NaN             4.0            4.0        4      9  2013
1272 2013-09-06 18:43:00  9/6/2013 18:43  ce28c2cc-7144-11e5-ba71-058fbc01cf0b  fd45d3ba-7144-11e5-ba71-058fbc01cf0b        Treat Warm-up    Communication   male      2012  Mixed  Mixed Breed/ Other/ I Don't Know         NaN      30        1.0         0.0   charmer             30         4            2.0           1  Wilmington    DE  19802      US       0               1  3/10/2014 16:18               2    NaN             5.0            5.0        4      9  2013

In [32]:
usertests2 = freestartusers.groupby("User_ID")["Created_at","Dog_ID"].count()
usertests2

Created_at  Dog_ID
User_ID                                                 
ce28c2cc-7144-11e5-ba71-058fbc01cf0b          78      78
ce28d776-7144-11e5-ba71-058fbc01cf0b           4       4
ce28e8c4-7144-11e5-ba71-058fbc01cf0b          20      20
ce28f2b0-7144-11e5-ba71-058fbc01cf0b          33      33
ce29002a-7144-11e5-ba71-058fbc01cf0b           6       6
...                                          ...     ...
ce9affb8-7144-11e5-ba71-058fbc01cf0b           2       2
ce9b001c-7144-11e5-ba71-058fbc01cf0b           4       4
ce9b01f2-7144-11e5-ba71-058fbc01cf0b           4       4
ce9b2eb6-7144-11e5-ba71-058fbc01cf0b           1       1
ce9b33b6-7144-11e5-ba71-058fbc01cf0b           4       4

[4233 rows x 2 columns]

In [33]:
usertests2["Dog_ID"].sum()

22356

In [34]:
usertests2["Dog_ID"].value_counts()

4     1436
7      845
2      643
1      575
6      142
3      141
20     133
5       91
11      51
16      23
14      17
8       14
15      13
40      10
12       8
10       8
23       7
13       7
25       7
18       5
19       5
44       4
33       4
17       4
26       3
28       3
31       3
34       3
9        3
27       3
22       2
24       2
41       2
42       2
35       2
36       1
45       1
76       1
21       1
29       1
93       1
43       1
46       1
50       1
78       1
39       1
59       1
Name: Dog_ID, dtype: int64

In [35]:
print("Percentage:" , 133/22356 * 100)

Percentage: 0.5949185900876722


#### Python code done by Dennis Lam